### PR TITLE
Improve `Nullable<T>` conversions & support

### PR DESCRIPTION
**Pull request by Ilia-Kosenkov**
- Add `TryFrom<RObj>` and its variations for `Nullable<T>`
- Add conversions to  and from `Options<T>`
- Add `map()` method that works similar to that of `Options<T>`, allowing acting on the state while propagating `NULL`
- Add a few doctests
- Add integration test
- Fix integration tests setup/config, update with newest versions of `{rextendr}` and `{roxygen2}`

### CODE REVIEW

File: extendr-api/src/wrapper/nullable.rs [hunk 1]
```diff
@@ -77,3 +131,88 @@ where
         }
     }
 }
+
+impl<T> Into<Option<T>> for Nullable<T> {
```
Comment (multimeric) [inline extendr-api/src/wrapper/nullable.rs:135]: I'm worried that we're using `Optional<T>` to represent a `T` that may be `NA` elsewhere in extendr. Should we consider adding a counterpart to this `Nullable<T>` that's like `NAable<T>` so we can drop the general `Option` conversion?

File: extendr-api/src/wrapper/nullable.rs [hunk 2]
```diff
@@ -77,3 +131,88 @@ where
...
+    fn into(self) -> Option<T> {
+        match self {
+            Nullable::NotNull(value) => Some(value),
+            _ => None,
+        }
+    }
+}
+
+impl<'a, T> Into<Option<&'a T>> for &'a Nullable<T> {
+    /// Convert a Nullable type into Option
```
Comment (multimeric) [inline extendr-api/src/wrapper/nullable.rs:153]: Clarify that this is for a reference

File: extendr-api/src/wrapper/nullable.rs (hunk 1 above)
Comment (multimeric) [inline extendr-api/src/wrapper/nullable.rs:135]: Why are we using `Into` here rather than `From`?

File: tests/extendrtests/DESCRIPTION [hunk 3]
```diff
@@ -1,23 +1,25 @@
...
 Authors@R: c(
     person("Andy", "Thomason", , "andy@andythomason.com", c("aut")),
     person("Claus O.", "Wilke", , "wilke@austin.utexas.edu", c("aut", "cre"),
-      comment = c(ORCID = "0000-0002-7470-9261"))
+      comment = c(ORCID = "0000-0002-7470-9261")),
+    person("Ilia", "Kosenkov", , "ilia.kosenkov@outlook.com", c("aut"), comment = c(ORCID = "0000-0001-5563-7840"))
   )
 Description:
-  This package provides tests for extendr functionality used from
+  Provides tests for extendr functionality used from
```
Comment (multimeric) [inline tests/extendrtests/DESCRIPTION:11]: These R changes seem unrelated to the PR?

File: extendr-api/src/wrapper/nullable.rs [hunk 4]
```diff
@@ -77,3 +131,88 @@ where
...
+    /// ```
+    /// use extendr_api::prelude::*;
+    /// test! {
+    ///     assert_eq!(Into::<Option<&i32>>::into(&Nullable::Null), None);
+    ///     assert_eq!(Into::<Option<&i32>>::into(&Nullable::NotNull(42)), Some(&42));
+    /// }
+    /// ```
+    fn into(self) -> Option<&'a T> {
+        match self {
```
Comment (multimeric) [inline extendr-api/src/wrapper/nullable.rs:162]: There's some redundancy here. Should we either a) delegate some of these impls: https://stackoverflow.com/a/56567737/2148718 or b) solve it with a macro (https://github.com/auto-impl-rs/auto_impl/). 

Because you haven't considered `mut` references here either.

File: extendr-api/src/wrapper/nullable.rs (hunk 1 above)
Comment (Ilia-Kosenkov) [reply extendr-api/src/wrapper/nullable.rs:135]: A valid point. However, due to the lack of docs, people just prefer to use known types, like `Option<i32>` not `Rint`. I am honestly OK with dropping options conversion at all, but it would be nice to have more methods to work with nullable type, like `map`, `and_then`, etc.

File: tests/extendrtests/DESCRIPTION (hunk 3 above)
Comment (Ilia-Kosenkov) [reply tests/extendrtests/DESCRIPTION:11]: I cleaned a little bit integration test. I can of course take it out into a separate PR if you feel like it should be done.

File: extendr-api/src/wrapper/nullable.rs (hunk 1 above)
Comment (Ilia-Kosenkov) [reply extendr-api/src/wrapper/nullable.rs:135]: No particular reason other than me not knowing the convention/best practices :)

File: extendr-api/src/wrapper/nullable.rs (hunk 4 above)
Comment (Ilia-Kosenkov) [reply extendr-api/src/wrapper/nullable.rs:162]: Maybe it is an argument for dropping `Option` here altogether and instead borrowing some methods from `Option` itself?

File: tests/extendrtests/DESCRIPTION (hunk 3 above)
Comment (Ilia-Kosenkov) [reply tests/extendrtests/DESCRIPTION:11]: Ok, I moved these bits to #447. As soon as it is merged, I'll drop some of the commits from this branch.

File: extendr-api/src/wrapper/nullable.rs (hunk 1 above)
Comment (multimeric) [reply extendr-api/src/wrapper/nullable.rs:135]: I think conversion from `Nullable` to `Option` is fine and idiomatic Rust. I just think we shouldn't have an automatic conversion from `Robj` to `Option<T>` specifically. Rather we should have an intermediate `NAable`. But this is probably a discussion for elsewhere.

Comment (multimeric) [reply extendr-api/src/wrapper/nullable.rs:135]: Yep, always use `From` instead of `Into` if you can. There are some scenarios where you can't do that though.

File: tests/extendrtests/DESCRIPTION (hunk 3 above)
Comment (multimeric) [reply tests/extendrtests/DESCRIPTION:11]: Once it's merged this part of the diff should disappear right?

File: extendr-api/src/wrapper/nullable.rs (hunk 4 above)
Comment (multimeric) [reply extendr-api/src/wrapper/nullable.rs:162]: Interesting idea, but I don't think that would be any cleaner. I think that might actually result in more code and redundancy rather than less.

File: extendr-api/src/wrapper/nullable.rs (hunk 1 above)
Comment (multimeric) [reply extendr-api/src/wrapper/nullable.rs:135]: Actually on second thoughts, it's not as confusing as it seems. When we have a vector, a given element can be `NA`, but a given element can't be `NULL`. So there's never a point at which there is confusion between `NA` and `NULL`. 

For example when we convert a `Robj` to `Optional<Vector<i32>>` the "optional" part is very clearly "is this null or not", because being `NULL` means there is no vector at all. If we convert `Robj` to `Vector<Optional<i32>>` the optional part is "is this particular element `NA`". So as long as this is documented, I think it's fine.

Do you agree?

Comment (Ilia-Kosenkov) [reply extendr-api/src/wrapper/nullable.rs:135]: Should we disallow conversion from `Option` to `Nullable` then? In that case, `Option` might represent `NA`-aware value.

Comment (multimeric) [reply extendr-api/src/wrapper/nullable.rs:135]: No, I don't think so. There shouldn't ever be overlap between `Option<T>` derived from `Nullable` and `Option<T>` derived from `NA`, so it's not confusing.

File: extendr-api/src/wrapper/nullable.rs (hunk 4 above)
Comment (multimeric) [reply extendr-api/src/wrapper/nullable.rs]: I agree that it doesn't make much sense to take a `&mut Nullable<T>` because there's no actual mutable state there. But on the other hand, there's no reason we can't convert that into an `Option`. It's just a reference with my permissions than usual.

Comment (Ilia-Kosenkov) [reply extendr-api/src/wrapper/nullable.rs]: So, right now we have no constraints on `T` when implementing `Nulalble<T> <--> Option<T>`. In reality though, `Nullable<T>` can be only constructed from something that can be `TryFrom<RObj>` or `Into<Robj>` (or some variation). I'd say we could put some trait constraints on `T` everywhere (something that says `T` should be `Robj` - compatible), in which case I see no real reason to expose `mut` reference support, since there is no scenario where we allow such things.

Comment (multimeric) [reply extendr-api/src/wrapper/nullable.rs]: Fair enough, there is no practical use case for converting mutable references.

But on that note, I like your idea for locking down the `impl` for `Nullable` using trait bounds to enforce that it's always a `Nullable<T> where T: TryFrom<Robj>`. I like the clear message of putting the trait bounds on the actual `enum` itself, but because https://github.com/rust-lang/rust/issues/44491 is still open, I guess it's best to just put the bounds on the `impl`.

Comment (Ilia-Kosenkov) [reply extendr-api/src/wrapper/nullable.rs]: So if you convert from `Nullable<T>` to `Option<T>`, we should require `T : TryFrom<Robj>`, it is a scenario for accepting a `Nullable<T>`. However, if we convert from `Option<T>` to `Nullable<T>`, we require `Into<Robj>` since it is a scenario of returning a `Nullable<T>`. The majority of types implement both traits, but not e.g. `ndarray.`
So we should be able to go `Nullable<ArrayView>` -> `Option<ArrayView>`, and then take an `Option<ArrayBase>` and convert to `Nullable<ArrayBase>` to return it back as nullable.

**Review by CGMossa — APPROVED** [review]

**Comment by multimeric** [comment]
Does this PR relate to https://github.com/extendr/extendr/issues/313 in any way?

**Comment by Ilia-Kosenkov** [comment]
Not directly, but maybe it covers some of the cases listed there. The original issue is about `Option<T>`, which we reserve for `NA`.
